### PR TITLE
✨ Implement read-only and write-only fields

### LINKED ISSUE #27: Read-Only Fields
I first want to say, I think this crate is great!

I am currently using it for OS development. When writing drivers for some x86 devices, the memory mapped registers used to control the devices sometimes contain read only bits. When programming the device, the register must be read in, other bits may be modified and then the register is written back (as not to modify the read only bits).

It would be useful to me, to be able to mark some members as read only (for example through `[bits(2, read_only = true)]`, where `read_only` would be false by default for backwards compatibility.

I think I could implement that myself (add a field to `Member` and modify `impl ToToken for Member` I guess) but wanted to ask first

### PR DESCRIPTION
**Pull request by paul1999**
Fixes #27

### CODE REVIEW

**Comment by wrenger** [comment]
Hi @paul1999 thanks a lot for your contribution.
I finally have the time to look into this and review your changes.

File: src/lib.rs [hunk 1]
```diff
@@ -504,31 +530,39 @@ impl Member {
...
 
     fn debug(&self) -> TokenStream {
-        if let Some(inner) = &self.inner {
-            let ident_str = inner.ident.to_string();
-            let ident = &inner.ident;
-            quote!(.field(#ident_str, &self.#ident()))
-        } else {
-            quote!()
+        match &self.inner {
+            Some(inner) if matches!(self.access, AccessMode::ReadWrite | AccessMode::ReadOnly) => {
```
Comment (wrenger) [inline src/lib.rs:541]: If you already have a match here, you could just match over both inner and access like so:

```rs
match (&self.inner, self.access) {
    (Some(inner), AccessMode::ReadWrite | AccessMode::ReadOnly) => { /*...*/ }
    _ => quote!(),
}
```

File: src/lib.rs [hunk 2]
```diff
@@ -504,31 +530,39 @@ impl Member {
...
-            let ident = &inner.ident;
-            let with_ident = format_ident!("with_{ident}");
-            quote!(this = this.#with_ident(#default);)
-        } else {
-            let offset = self.offset;
-            let base_ty = &self.base_ty;
-            quote!(this.0 |= (#default as #base_ty) << #offset;)
+
+        match &self.inner {
+            Some(inner) if matches!(self.access, AccessMode::ReadWrite | AccessMode::WriteOnly) => {
```
Comment (wrenger) [inline src/lib.rs:556]: Same here

File: src/lib.rs [hunk 3]
```diff
@@ -711,6 +769,14 @@ fn parse_field(
                 ));
             }
 
+            // Ensure padding fields remain inaccesible
+            if ignore && access.filter(|mode| *mode != AccessMode::None).is_some() {
+                return Err(syn::Error::new(
```
Comment (wrenger) [inline src/lib.rs:774]: `access.is_some_and(|a| a != AccessMode::None)` might be a little bit more clear than filter.

File: src/lib.rs [hunk 4]
```diff
@@ -798,6 +871,18 @@ impl Parse for BitsAttr {
                     attr.into = Some(input.parse()?);
                 } else if ident == "from" {
                     attr.from = Some(input.parse()?);
+                } else if ident == "access" {
+                    let mode = input.parse::<Ident>()?;
+
+                    if mode == "RW" {
```
Comment (wrenger) [inline src/lib.rs:877]: This function becomes quite complex.
What if we implement Parse for AccessMode and put this parsing code there?
Also, we might want to have a meaningful error if the attribute has an invalid value (not just silently ignore it).

**Review by wrenger — COMMENTED** [review]
Your additions are quite excellent, not much to criticize. The first three comments are minor style suggestions (that are rather optional). Though, the last one about a proper error for invalid access values might be more important.

File: src/lib.rs (hunk 3 above)
Comment (paul1999) [reply src/lib.rs:766]: There are so many methods on `Option` and `Result`, I always confuse them :D

File: src/lib.rs (hunk 4 above)
Comment (paul1999) [reply src/lib.rs]: Thought about using an extra Parse implementation aswell...the fact I missed giving an error message here convinced me enough to do it.

**Comment by wrenger** [comment]
Thanks, I merged it and released it with 0.5.6.

**Pull request merged**